pkg/gofr/service/ratelimiter.go

```diff
@@ -0,0 +1,475 @@
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/time/rate"
+)
+
+var (
+	errQueueFull      = errors.New("request queue is full")
+	errShuttingDown   = errors.New("rate limiter is shutting down")
+	errPanicRecovered = errors.New("panic recovered in request processing")
+)
+
+// RateLimiterConfig holds configuration for creating a new rate limiter.
+type RateLimiterConfig struct {
+	Limit, MaxQueue int
+	Duration        time.Duration
+}
+
+// AddOption creates a new rate limiter with the config settings and wraps the provided HTTP client.
+func (r *RateLimiterConfig) AddOption(h HTTP) HTTP {
+	if hs, ok := h.(*httpService); ok {
+		rl := NewRateLimiter(r.Limit, r.Duration, r.MaxQueue)
+		rl.HTTP = h
+		rl.Logger = hs.Logger
+
+		return rl
+	}
+
+	if _, ok := h.(*RateLimiter); ok {
+		return h
+	}
+
+	rl := NewRateLimiter(r.Limit, r.Duration, r.MaxQueue)
+	rl.HTTP = h
+
+	return rl
+}
+
+type RateLimiter struct {
+	HTTP
+	limiter        *rate.Limiter
+	requests       chan requestWrapper
+	maxQueue       int
+	mutex          sync.Mutex
+	done           chan struct{}
+	wg             sync.WaitGroup
+	isShuttingDown bool
+	currentQueue   int32
+	Logger
+}
+
+type requestWrapper struct {
+	execute func() (*http.Response, error)
+	respCh  chan *requestResponse
+	ctx     context.Context
+	cancel  context.CancelFunc
+}
+
+type requestResponse struct {
+	response *http.Response
+	err      error
+}
+
+// NewRateLimiter creates a rate limiter.
+func NewRateLimiter(limit int, window time.Duration, maxQueue int) *RateLimiter {
+	if maxQueue <= 0 {
+		maxQueue = 1000
+	}
+
+	lim := rate.NewLimiter(rate.Every(window/time.Duration(limit)), limit)
+	rl := &RateLimiter{
+		limiter:      lim,
+		requests:     make(chan requestWrapper, maxQueue),
+		maxQueue:     maxQueue,
+		done:         make(chan struct{}),
+		currentQueue: 0,
+	}
+
+	rl.Start()
+
+	return rl
+}
```

Comment:
I would have used a context instead of a channel

```suggestion
func NewRateLimiter(ctx context.Context,limit int, window time.Duration, maxQueue int) *RateLimiter {
	if maxQueue <= 0 {
		maxQueue = 1000
	}

	lim := rate.NewLimiter(rate.Every(window/time.Duration(limit)), limit)
	rl := &RateLimiter{
		limiter:      lim,
		requests:     make(chan requestWrapper, maxQueue),
		maxQueue:     maxQueue,
		currentQueue: 0,
	}

	rl.Start(ctx)

	return rl
}
```

And Start would have wait for ctx.Done(), and not r.done